components/region_cache_memory_engine/src/keys.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp;
+
+use bytes::{BufMut, Bytes, BytesMut};
+use skiplist_rs::KeyComparator;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ValueType {
+    Deletion = 0,
+    Value = 1,
+}
+
+// See `compare` of InternalKeyComparator, for the same user key and same
+// sequence number, ValueType::Value is less than ValueType::Deletion
+pub const VALUE_TYPE_FOR_SEEK: ValueType = ValueType::Value;
+pub const VALUE_TYPE_FOR_SEEK_FOR_PREV: ValueType = ValueType::Deletion;
+
+impl TryFrom<u8> for ValueType {
+    type Error = String;
+    fn try_from(value: u8) -> std::prelude::v1::Result<Self, Self::Error> {
+        match value {
+            0 => Ok(ValueType::Deletion),
+            1 => Ok(ValueType::Value),
+            _ => Err(format!("invalid value: {}", value)),
+        }
+    }
+}
+
+pub struct InternalKey<'a> {
+    pub user_key: &'a [u8],
+    pub v_type: ValueType,
+    pub sequence: u64,
+}
+
+const ENC_KEY_SEQ_LENGTH: usize = std::mem::size_of::<u64>();
+
+impl<'a> From<&'a [u8]> for InternalKey<'a> {
+    fn from(encoded_key: &'a [u8]) -> Self {
+        decode_key(encoded_key)
+    }
+}
+
+#[inline]
+pub fn decode_key(encoded_key: &[u8]) -> InternalKey<'_> {
+    assert!(encoded_key.len() >= ENC_KEY_SEQ_LENGTH);
+    let seq_offset = encoded_key.len() - ENC_KEY_SEQ_LENGTH;
+    let num = u64::from_be_bytes(
+        encoded_key[seq_offset..seq_offset + ENC_KEY_SEQ_LENGTH]
+            .try_into()
+            .unwrap(),
+    );
+    let sequence = num >> 8;
+    let v_type = ((num & 0xff) as u8).try_into().unwrap();
+    InternalKey {
+        user_key: &encoded_key[..seq_offset],
+        v_type,
+        sequence,
+    }
+}
+
+#[inline]
+pub fn extract_user_key_and_suffix_u64(encoded_key: &[u8]) -> (&[u8], u64) {
+    assert!(encoded_key.len() >= ENC_KEY_SEQ_LENGTH);
+    let seq_offset = encoded_key.len() - ENC_KEY_SEQ_LENGTH;
+    let num = u64::from_be_bytes(
+        encoded_key[seq_offset..seq_offset + ENC_KEY_SEQ_LENGTH]
+            .try_into()
+            .unwrap(),
+    );
+
+    (&encoded_key[..seq_offset], num)
+}
+
+/// Format for an internal key (used by the skip list.)
+/// ```
+/// contents:      key of size n     | value type | sequence number shifted by 8 bits
+/// byte position:         0 ..  n-1 | n          |  n + 1 .. n + 7
```

Comment:
The pattern of InternalKey should be re-corrected with the following format?

```suggestion
/// contents:          key of size n | sequence number shifted by 8 bits | value type 
/// byte position:       0 ..  n - 1 |          n .. n + 6               | n + 7
```

Reply:
The last u64 is encoded as Big-endian.

Reply:
Got.

IMO, it's better to add the annotations that the last `seq_num` with `val_type` is encoded with Big-endian.